Docs/fixed-folders (#1163)
MaibornWolff/codecharta: gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md

---
categories:
    - How-to
title: Fixate folders using a custom cc.json
---

# Introduction

The [Squarified-Tree-Map-Algorithm](https://www.win.tue.nl/~vanwijk/stm.pdf) generates a layout for the visualized map.
The layout varies depending on the area of each node. Adding a new file to a folder might increase the area of the parent folder,
so that the algorithm decides that the parent-folder needs to be relocated. Another way to follow this problem is by changing the area-metric.

> Notice, that most folders will not be located at the same position as before.

Reviewer: ```suggestion
Thus, folders may not be located at the same position anymore as before.
```

Ideally move the sentence before the `Another way to follow...`

Setting this property is restricted to the top level folders and won't have any effect on sub-folders (folders in folders).

# API

```json
 "fixedPosition": {
    "left": 60,
    "top:": 40,
    "width": 35,
    "height": 55
 }
```

The property values must be numbers in the range between 0 and 100. They represent the size of the value in percent.

## Example

```json
{
	"projectName": "example-project",
	"apiVersion": "1.2",
	"nodes": [
		{
			"name": "root",
			"type": "Folder",
			"attributes": {},
			"children": [
				{
					"name": "folder_1",
					"type": "Folder",
					"attributes": {},
					"children": [
						{
							"name": "children_1",
							"type": "File",
							"attributes": {
								"custom_metric": 2
							},
							"children": []
						}
					],
					"fixedPosition": {
						"left": 5,
						"top:": 5,
						"width": 50,
						"height": 30
					}
				},
				{
					"name": "folder_2",
					"type": "Folder",
					"attributes": {},
					"children": [],
					"fixedPosition": {
						"left": 60,
						"top:": 40,
						"width": 35,
						"height": 55
					}
				}
			]
		}
	]
}
```

> Since a margin between folders is recommended, we set the coordinates for the folder in the top-left corner
> to `left: 5` and `top: 5`. The margin between the border of the map and `folder_1` is therefore `5`, but can be chosen at will.
> In order to set the coordinates to adjacent folders, we can again apply any margin between `folder_1` and `folder_2`. We chose
> a margin of `5` again, so that the coordinates of `folder_2` must be `left: 60` and `top: 40`.

Reviewer: ```suggestion
A margin between folders is recommended. To apply one, set the coordinates for the folder in the top-left corner
to e.g., `left: 5` and `top: 5`. The margin between the border of the map and `folder_1` is therefore `5`. It may be chosen at will.
In order to define the coordinates of adjacent folders, apply any margin between `folder_1` and `folder_2`. The example uses a margin of `5`, so that the coordinates of `folder_2` must be `left: 60` and `top: 40`.
```

## Restrictions

![Example]({{site.baseurl}}/assets/images/posts/how-to/fixate-folders/fixate-folder-example.jpg)

> In order to build a valid custom `.cc.json`, these rules must be followed:

Reviewer: ```suggestion
The following rules apply in order to build a valid custom `.cc.json`:
```

-   The values of `left`, `top`, `width` and `height` must be in range of `[0, 100]
-   The value `left + width` or `top + height` must be in a range of `[0, 100]`.
-   Folders may not overlap.
-   Leaving space between folders (for visibility reasons) is recommended.

> -   Only folders on the root can have the `fixedPosition`-attribute.
> -   If there exists at least on fixated folder, all folders on the root must be fixated.

Reviewer: ```suggestion
-   All folders on the root level have to use the `fixedPosition` attribute.
```

Author: `on the root` still sounds misleading to me.
What about:
```suggestion
-   All direct sub-folders of the root have to use the `fixedPosition` attribute.
```

Reviewer: ```suggestion
-   All root folders have to use the `fixedPosition` attribute.
```

What about keeping it brief?

Author: There is only one root folder and it should not have the `fixedPosition` attribute. All direct children of that folder require it though, if at least one of them has it.

Reviewer: ```suggestion
-   All children of the root folder require the `fixedPosition` attribute.
```

How about that? (I explicitly did not mention "if at least one of them has the attribute set". This whole section is only about the fixed position)

---
categories:
    - How-to
title: Fixate folders using a custom cc.json
---

# Introduction

The [Squarified-Tree-Map-Algorithm](https://www.win.tue.nl/~vanwijk/stm.pdf) generates a layout for the visualized map.
The layout varies depending on the area of each node. Adding a new file to a folder might increase the area of the parent folder,
so that the algorithm decides that the parent-folder needs to be relocated. Another way to follow this problem is by changing the area-metric.
Notice, that most folders will not be located at the same position as before.

# Fixate folders

> In order to prevent folders from changing locations, we introduced a new attribute to the `.cc.json` called `fixed`.

Author: I just talked about the same thing with @Richargh 
We should either go with `fixedPosition` or `fixedLocation`. Position is more suiting imo.

This attribute will not be generated during the analysis. The user needs to edit the `.cc.json` by hand.

# API

```json
 "fixed": {
    "x": 60,

>     "y:": 40,

Author: `left` and `top` sounds great.

    "width": 35,
    "height": 55
 }
```

The `fixed`-attribute includes a coordinate `x` and `y`. These represent the top-left corner of the folder. `width`

> defines the length in x-direction and `height` defines the length in y-direction.

Reviewer: We might not need this if we rename the properties to something more intuitive?

Author: True. Should we keep `width` and `height` or should we use `width` and `length`?

Reviewer: We should keep `height`.

> Note, that these numbers represent percentages. The overall map size is therefore `100`.

Reviewer: ```suggestion
The property values must be integers in the range between 0 and 100. They represent the size of the value in percent.
```

Author: ```suggestion
The property values must be numbers in the range between 0 and 100. They represent the size of the value in percent.
```

You can also specify `x: 1.5`

Reviewer: Ha, alright.

## Example

```json
{
	"projectName": "example-project",
	"apiVersion": "1.2",
	"nodes": [
		{
			"name": "root",
			"type": "Folder",
			"attributes": {},
			"children": [
				{
					"name": "folder_1",
					"type": "Folder",
					"attributes": {},
					"children": [
						{
							"name": "children_1",
							"type": "File",
							"attributes": {
								"custom_metric": 2
							},
							"children": []
						}
					],
					"fixed": {
						"x": 5,
						"y:": 5,
						"width": 50,
						"height": 30
					}
				},
				{
					"name": "folder_2",
					"type": "Folder",
					"attributes": {},
					"children": [],
					"fixed": {
						"x": 60,
						"y:": 40,
						"width": 35,
						"height": 55
					}
				}
			]
		}
	]
}
```

> Notice, that `folder_1` start at `x:5`, `y:5`, because we apply a margin of `5` around the map.

Reviewer: Is the margin always present and is it fixed at the value 5?

Author: No. This is really confusing here. I will rephrase this again.

`folder_2` starts at `x:60`, `y:40`, because `folder_1` ends at `x:55`, `y:35` and we apply another margin of `5` between those.

## Restrictions

![Example]({{site.baseurl}}/assets/images/posts/how-to/fixate-folders/fixate-folder-example.jpg)

In order to build a valid custom `.cc.json`, these rules must be followed:

-   The value `x + width` or `y + height` must be in a range of `[0, 100]`.

> -   Folders can't overlap. Applying a margin between folders is recommended.

Reviewer: The margin can not be changed by the user, right? If that's the case, the second sentence seems to be redundant?

Author: The margin is determined by the space that is left over after you defined your `fixed`-object. The margin can be different on every side.

Maybe `Leaving some space between folders for visibility reasons is recomended.` is better?

Reviewer: Smart. Even though we can't determine all bad input cases due to that. Seems like it's a well made consideration.

Reviewer: Your suggestion sound pretty good. I'd go with

```suggestion
-   Folders may not overlap.
-   Leaving space between folders (for visibility reasons) is recomended.
```
